Finish modularization, and lots of refactoring

jvanstraten opened the pull request:
Followup on #364

## In this PR

 - [x] Removal of unused code as agreed upon
    - [x] Remove CC-light-specific code and passes
    - [x] Remove defunct passes

 - [x] General cleanup on a per-file basis
    - Add structure to file organization (see #404)
    - Apply naming conventions to all classes according to [the style guide](https://github.com/QE-Lab/OpenQL/blob/develop/CONTRIBUTING.md)
    - Refactor all raw pointers/`new` allocations to smart pointers to vastly reduce memory usage and prevent segfaults
    - Files (old location, strikethrough means file was deleted in full or for the most part):
       - [x] ~~`buffer_insertion.[cc|h]`~~
       - [x] `circuit.[cc|h]`
       - [x] `classical.[cc|h]`
       - [x] `clifford.[cc|h]`
       - [x] ~~`commute_variation.[cc|h]`~~
       - [x] ~~`compiler.[cc|h]`~~
       - [x] ~~`decompose_toffoli.[cc|h]`~~
       - [x] ~~`eqasm_compiler.[cc|h]`~~
       - [x] `gate.[cc|h]`
       - [x] `hardware_configuration.[cc|h]`
       - [x] `interactionMatrix.[cc|h]`
       - [x] `ir.[cc|h]`
       - [x] `kernel.[cc|h]`
       - [x] ~~`latency_compensation.[cc|h]`~~
       - [x] `mapper.[cc|h]`
          - [x] Extract qubit grid utilities and make part of platform
          - [x] Extract virt2real utilities and make part of common
          - [x] Extract initial placement code for separate pass
          - [x] Make Options class for storing parsed options, instead of the mapper referring to global options all over
          - [x] Extract and refactor the FreeCycle class
          - [x] Extract and refactor the Past class
          - [x] Extract and refactor the Alter class
          - [x] Extract and refactor the Future class
          - [x] Refactor the main Mapper class
       - [x] ~~`matrix.h`~~
       - [x] ~~`metrics.[cc|h]`~~
       - [x] `openql.h`
       - [x] `openql_i.[cc|h]`
       - [x] ~~`optimizer.[cc|h]`~~
       - [x] `options.[cc|h]`
       - [x] `platform.[cc|h]`
       - [x] `program.[cc|h]`
       - [x] `ql_config.h`
       - [x] ~~`qsoverlay.[cc|h]`~~
       - [x] `report.[cc|h]`
       - [x] `resource_manager.[cc|h]`
       - [x] `scheduler.[cc|h]`
       - [x] `unitary.[cc|h]`
       - [x] `visualizer.h`
       - [x] `visualizer_cimg.[cc|h]`
       - [x] `visualizer_circuit.[cc|h]`
       - [x] `visualizer_common.[cc|h]`
       - [x] `visualizer_interaction.[cc|h]`
       - [x] `visualizer_mapping.[cc|h]`
       - [x] `visualizer_types.h`
       - [x] `write_sweep_points.[cc|h]`
       - [x] ~~`arch/cc_light/cc_light_eqasm.[cc|h]`~~
       - [x] ~~`arch/cc_light/cc_light_eqasm_compiler.[cc|h]`~~
       - [x] `arch/cc_light/cc_light_resource_manager.[cc|h]`
       - [x] `arch/cc/backend_cc.[cc|h]`
       - [x] `arch/cc/bundle_info.h`
       - [x] `arch/cc/codegen_cc.[cc|h]`
       - [x] `arch/cc/datapath_cc.[cc|h]`
       - [x] `arch/cc/options_cc.h`
       - [x] `arch/cc/settings_cc.[cc|h]`
       - [x] `arch/cc/types_cc.h`
       - [x] `arch/cc/vcd_cc.[cc|h]`
       - [x] `cqasm/cqasm_reader.[cc|h]`
       - [x] ~~`passes/passes.[cc|h]`~~
       - [x] ~~`passes/passmanager.[cc|h]`~~

 - [x] Basic IR restructuring
    - [x] Use `tree-gen` base and reference classes for IR, to smoothen transition to generated tree (if generating it isn't more trouble than it's worth)
    - [x] Add creg and breg count to platform
    - [x] Deduplicate information as much as possible at this stage, and actually check the required relations where duplication remains (although everything is still public right now, so anyone who really wants to mess things up still can)
    - [x] Properly document all the fields used in the IR structures
       - [x] Program
       - [x] Kernel
       - ~~Gate~~
       - [x] Bundle
    - [x] Properly document all the methods used in the IR structures
       - [x] Program
       - ~~Kernel~~
       - ~~Gate~~
       - ~~Classical~~
       - [x] Bundle
    - (note: it's not really worth properly documenting stuff that's on the todo list to be revised completely in the near future, so I took this out to get this merged quicker)
    - [x] Remove unused fields and turn things only used by a particular pass into annotations
       - [x] Annotation for pass-specific statistics data
       - [x] Annotation for sweep points
       - [x] Remove matrices

 - [x] Basic platform restructuring
    - [x] Properly document all the fields used in the platform structure
    - [x] Abuse `"eqasm_compiler"` key for specifying compiler configuration, either via filename or in-place
    - [x] Add built-in configuration files, so people can stop having to refer to the `tests` directory for them
    - [x] Introduce architectures and architecture variants for the above
    - [x] Allow architecture-specific code to preprocess the platform for additional validity checks, desugaring, etc.
    - [x] Place default backend pass construction code where it belongs (`arch.*`)
    - [x] Place the default platform configuration files where they belong: in `src/arch/*`, baked into the OpenQL library via resource generation/
    - [x] Add APIs for modifying platform configuration data, by adding methods for getting the configuration file defaults from OpenQL as Python object representation of the JSON structure, and methods to construct a platform from that directly (instead of indirectly by writing the result to a file first)
    - [x] Add simplified platform constructors that use the provided defaults for platform configuration, and thus don't need a JSON file or JSON data
    - [x] Add tests for the above
    - [x] Document it
       - [x] In-code documentation
       - [x] Runtime documentation for architectures
          - [x] `cc`
          - [x] `cc_light`
          - [x] `none`
       - [x] Runtime documentation for platform configuration files
       - [x] Online documentation (generated from runtime documentation)

 - [x] Revise/future-proof pass management
    - [x] Remove `eqasm_compiler` class (no difference now between it and regular passes)
    - [x] Rebuild the pass manager (see #404)
       - [x] Base class
       - [x] Derived classes with shared functionality (analyzer vs transformer etc)
       - [x] Pass groups
       - [x] Conditional/looping pass groups
       - [x] Pass factory and pass registration (hardcoded for now; code generator postponed)
       - [x] Pass manager toplevel
       - [x] JSON configuration file parser
    - [x] Port passes from the old pass manager to the new one
       - [x] Visualizer (basically done, just need to look up what they should do, test them, and add runtime docs; also does not comply with code style 100% yet, but that can easily be postponed)
       - [x] Statistics reporter
       - [x] cQASM writer
       - [x] cQASM reader
       - [x] Scheduler
       - [x] RC scheduler
       - [x] Placer (i.e. initial placement without routing, first part of mapper)
       - [x] Router (i.e. heuristic placement by means of adding swaps/moves as needed)
       - [x] Clifford optimizer
       - [x] CC backend
       - [x] Sweep point writer
    - [x] Replace old pass manager with the new one
       - [x] Add compatibility mode for global options
       - [x] Update compiler configuration files
       - [x] Make sure the tests pass
       - [x] Update/test/fix visualizer examples
    - [x] Add tests specifically for the new pass management system
    - [x] Document it
       - [x] In-code documentation
       - [x] Runtime documentation

 - [x] Revise/future-proof resource management
    - [x] Rebuild the resource manager
       - [x] Add `ClonablePtr` abstraction to prevent the need of `clone()` code duplication in the resources
       - [x] Remove `platform_resource_manager` class; all platforms can use all resource types, identified by name (possibly prefixed with platform type, just like pass management)
       - [x] Rewrite resource interface to use only one function with a commit flag versus available() and reserve() to prevent unnecessary code duplication, considering the implementation of these functions is almost identical
       - [x] Resource factory and pass registration (hardcoded for now; code generator postponed)
       - [x] Resource manager toplevel
       - [x] JSON configuration file/section parser
    - [x] Port resources from the old resource manager to the new one
       - [x] Temporary compatibility layer for testing
       - [x] Figure out how to generalize the resources beyond CC-light (and figure out non-CC-light-specific names for them) to make them more reusable for CC (and potentially other targets in the future)
       - [x] Port qubits resource
       - [x] Port QWG resource (using generic instrument resource)
       - [x] Port measurement units resource (using generic instrument resource)
       - [x] Port edge resource (using generic instrument resource)
       - [x] Port detuned qubits resource (using generic instrument resource)
       - [x] Port multi-core channel resource
    - [x] Replace old resource manager with the new one
       - [x] Make sure the tests pass
    - [x] Document it
       - [x] In-code documentation
       - [x] Runtime documentation

 - [x] Revise/future-proof option system
    - [x] Convert as many global options to pass options as possible, with compatibility layer that sets the defaults for the pass options based on the global options (done on-the-fly as I add passes)
    - [x] Properly document all the options (pass options and remaining/compatibility global options) using the new runtime documentation system

 - [x] Refactor/future-proof statistics/metrics to avoid code duplication and make it easier to add more later

 - [x] Update Python API
    - [x] `ql.Platform`
       - [x] Wrap not only the internal Platform, but also the compiler configuration, using a new optional constructor argument for the compiler configuration file, to provide complete backward-compatibility (as opposed to the current modularization interface, that modifies the object construction/function call flow)
       - [x] Add methods for querying runtime documentation for available pass types and pass options
       - [x] Add methods for querying runtime documentation for available resources
       - [x] Add methods for querying and modifying the pass list and pass options
    - [x] `ql.Program`
       - [x] Point `ql.Program.compile()` to new pass manager, based on the platform that the program was constructed with
    - [x] Add method for querying runtime documentation for available global options
    - [x] Documentation
       - [x] Monkey-patch the SWIG-generated Python function signature part of the docstrings to use Python type names and syntax, and to make it play nice together with Sphinx
       - [x] Write docs for misc functions
       - [x] Write docs for Platform
       - [x] Write docs for Program
       - [x] Write docs for Kernel
       - [x] Write docs for CReg
       - [x] Write docs for Operation
       - [x] Write docs for Unitary
       - [x] Write docs for Compiler
       - [x] Write docs for Pass
       - [x] Write docs for cQasmReader

 - [x] Documentation
    - [x] Revise ~~basically all~~ some of the hand-written stuff. Pretty sure everything is out of date and/or relies on internal knowledge of OpenQL's codebase.
       - [x] Introduction
       - [x] User manual and tutorials (at least an initial draft that's in sync with this version of OpenQL)
       - [x] Reference
          - [x] Python API
          - [x] C++ API
          - [x] Configuration
             - [x] Platform
             - [x] Compiler
             - [x] cQASM reader (described in reader pass docs)
             - [x] Visualizer (described in visualizer pass docs)
          - [x] Architectures
          - [x] Global options
          - [x] Passes
          - [x] Resources
       - [x] Developer documentation
       - [x] Colophon (changelog!)
       - [x] Move all pages not dealt with yet to an "old pages" section
    - [x] Build magic in Sphinx `conf.py` to generate online docs based on runtime docs (available passes, available resources, and available global/pass options)
    - [x] Format the runtime docs better to be properly readable in Python and to look okay in ReadTheDocs per above
    - [x] Update style guide with structure information

 - [x] Figure out where the virtual-to-real qubit mapping information for the visualizer is going (SwapParameters), since the visualizer doesn't see them for some reason

## Future refactoring/utility work (i.e. postponed for now)

 - More IR/platform restructuring
    - Field privatization: right now everything is public with no access checks whatsoever, making it very easy to make mistakes while adding/changing features
    - Merge common functionality into common functions, rather than everyone having to deal with everyone else's intricacies whenever the IR is mutated

 - Revise gate representation and construction
    - Separate gate instances from gate types
    - Add semantical information to those gate types (such as clifford gate index if any, rotation axis if any, etc) instead of trying to derive this from gate names everywhere, and replace builtin gates with this as well (the current polymorphic gate classes make zero sense, the polymorphism isn't actually used, just makes things more difficult, and somehow *increases* code duplication)
    - Add key in platform description that specifies the expected prototype for a gate, presenting the same information that libqasm needs to parse a cQASM file (again, so this doesn't need to be heuristically derived from gate names)
    - Instead of spamming different fields for every possible operand that a gate might have, homogenize to a single list of polymorphic operands
    - Tear gate decomposition out of `Kernel` and stick it in a configurable pass instead
    - Tear unitary decomposition out of `Kernel` and stick it in a pass instead
    - Move generative functions for kernels (such as inverting a pure-quantum kernel or generating a controlled subkernel) to a more suitable place, or remove them from OpenQL altogether

 - Wrap JSON loader/data structure
    - Design something to make it easy to read JSON structures without configuration errors invoking `CRASH_AND_BURN()`.
    - Retain line number information in the loaded structure, or alternatively track the hierarchy leading up to a faulty JSON object, so configuration errors provide more context for the user
    - Add errors (or at the very least warnings) for unknown keys, so misspellings of configuration keys are actually detected

 - Revise logging system
    - Introduce hierarchical sections into the log output, each with its own loglevel
    - Add profiling features (performance monitoring) to the above
    - Use a formatting library instead of the objectively terrible C++ stream operators
    - Make sure the messages printed are actually sane

 - Revise exception system
    - Separate exceptions representing internal compiler errors (with stack trace etc) from more user-friendly errors
    - Maybe add line number information to the above (as in the line number in the algorithm source rather than the line of C++ that threw the exception; this is absolutely not an easy task!)

 - Add unit testing framework to build system
    - ... and actually add some unit tests
    - ... or at least start writing them for new stuff

 - Distribute default configuration files with Python installation properly
    - ... so people don't have to copy them from the ****ing `tests` directory, and things stop lighting on fire outside of the context of the test suite

 - Revise representation of classical code/advanced IR restructuring (i.e. OpenQL implementation of cQASM 2.0)

jvanstraten commented:
Supported API use cases (still incomplete, work in progress)

## Use case 1: the old-fashioned way

The result must be the same as before the refactoring/pass management
rewrite, for as far as passes/options were not removed.

 - global options affecting compilation may be set any time from the start
   of the program
 - build platform with name + JSON filename; platform JSON file has
   recognized eqasm_compiler string or key is missing/empty
 - build program using name/platform
 - build kernel(s) using name/platform
 - add kernel(s) to program
 - global options may be set until this point to affect compilation
 - call Program.compile()

## Use case 2: no changes to Python, only to config files

Basically, the old-fashioned way within python, but with compiler configuration
specified in the platform, with compatibility mode enabled.

 - global options affecting compilation may be set any time from the start
   of the program, except for options that affect the pass order
   (clifford_* and prescheduler), which are ignored
 - build platform with name + JSON filename; platform JSON file either has:
    - an unrecognized eqasm_compiler string that is instead a filename
      pointing to a compiler configuration file (filename relative to the
      platform JSON file, or as fallback relative to the working
      directory); or
    - an object for eqasm_compiler with a strategy subkey, treated as an
      inline compiler configuration object
 - OR build platform with name + platform JSON filename + explicit compiler
   JSON filename; the given compiler JSON filename will override whatever
   is in the eqasm_compiler node
 - build program using name/platform
 - build kernel(s) using name/platform
 - add kernel(s) to program
 - global options may be set until this point to affect compilation
 - call Program.compile()

## Use case 3: using a mix of global and pass options within Python

i.e. use the new system for new features, but with as much backward
compatibility as possible.

 - global options affecting compilation may be set any time from the start
   of the program until specified, except for options that affect the pass
   order (clifford_* and prescheduler), which are ignored
 - build platform as in use case 1 or 2
 - at any point from here until Program.compile() or Compiler.compile(),
   call Platform.set_option(opt, val) (which is shorthand for
   Platform.get_compiler().set_option(opt, val); either is acceptable) to
   set pass options for the default or custom pass list; from this point
   onward, global options will cease to affect compilation (this includes
   output_dir, which is superseded by the **.output_prefix pass option).
   the other calls on the Compiler object returned by get_compile() can
   of course also be used, for instance to add or remove a pass
 - build program using name/platform
 - build kernel(s) using name/platform
 - add kernel(s) to program
 - global options may be set until this point to affect compilation
 - call Program.compile() or Compiler.compile(program)

## Use case 4: using pass management options only

 - all global options except those affecting gate creation are IGNORED by
   compilation, including output_dir (which is superseded by the
   **.output_prefix pass option)
 - build platform as in use case 1 or 2, but without compatibility_mode key
   (or with it but set to false)
 - at any point from here until Program.compile() or Compiler.compile(),
   call Platform.set_option(opt, val) (which is shorthand for
   Platform.get_compiler().set_option(opt, val); either is acceptable) to
   set pass options for the default or custom pass list. the other calls on
   the Compiler object returned by get_compile() can of course also be
   used, for instance to add or remove a pass
 - build program using name/platform
 - build kernel(s) using name/platform
 - add kernel(s) to program
 - global options may be set until this point to affect compilation
 - call Program.compile() or Compiler.compile(program)

## Use case 5: like 4, but with a more sane object flow (compiler object)

TODO

wvlothuizen reviewed (approved):
Really impressive, big improvement to architecture, implementation and documentation. Builds out of the box, on my Mac.

For the review, I focussed on the CC backend, and code/passes relevant to that. I found no real issues, just a few minor comments.

I also checked the PycQED tests, which all succeed after correcting a .JSON config  file (which contained errors that were not caught before).

I'd suggest to bump the version now already to distinct trial installations (which now get '0.8.1.dev5') from the former version .

jvanstraten commented on include/ql/ir/classical.h:
```diff
@@ -0,0 +1,89 @@
+namespace ir {
+
+enum class ClassicalOperationType {
+    ARITHMATIC, RELATIONAL, BITWISE
```
Fixed in d45042f

jvanstraten commented on src/ql/api/platform.cc:
```diff
@@ -0,0 +1,163 @@
+
+/**
+ * Constructs a platform. name is any name the user wants to give to the
+ * platform; it is only used for report messages. platform_config_file must
```
Fixed in 6bc7a5b

jvansomeren commented:
Clean build.
Clean docbuild.
Now reading documentation.

Python API: get_pass_types  retrieves documentation but name looks like something retrieving a handle on e.g. the passes: "get_...". Prefer that documentation retrieval gets other name prefixes such as "print", as used by later APIs.

jvanstraten commented:
> Python API: get_pass_types retrieves documentation but name looks like something retrieving a handle on e.g. the passes: "get_...". Prefer that documentation retrieval gets other name prefixes such as "print", as used by later APIs.

@jvansomeren 

While `print_*` versions of all documentation retrieval functions exist, just dumping to stdout is not always what you want. I would argue that the primitive function is always a getter that returns a string (which you can then print, or in the case of docs/conf.py convert to reStructuredText fragments), but for compatibility and consistency a `print_*` version should exist for all documentation getters as well.

That said, the name of that particular doc getter could indeed be changed to better express what it does, for example `[get|print]_pass_type_docs()`, otherwise it might be confused with something that just returns a list of available pass type names (for which there is no API right now, though one could easily be added). Alternatively I could replace the `get_` prefix with for example `dump_` for all documentation getters. Thoughts?

jvanstraten commented:
@wvlothuizen

Wheels built by CI according to e426c1e (Python 3.9) and 933163e (Python 3.6 through 3.8) should now be available for further testing via `pip3 install qutechopenql==0.8.1.dev6` for Windows, MacOS, and Linux.

jvansomeren commented:
I would then propose to prefix the string returning getters with dump_ instead of get_.

jvansomeren commented:
In Python API, compilation strategy is mentioned, a default compilation strategy which not be good enough and that you can change it. Only in class Platform it is described (with the compiler config) while it should be described in class Compiler at least.

jvansomeren commented:
What is the use of pass groups? One can make them, set options to subpasses. There are so many interfaces but it is not clear when reading them why they are there.

jvanstraten commented:
Some use cases I envisioned for pass groups, off the top of my head:

 - Help organize/hide the complexity of a bunch of passes that basically always need to follow one another to achieve a functional compiler, like the many subpasses of the CC-light backend. In this case I envision a sort of meta-pass that generates into a group (upon a call to `construct()`; the `on_construct()` virtual function allows for this behavior), at which point the group APIs would be useful if you're a power user and want to change things anyway, or want to insert an analysis pass halfway or something.
 - The hierarchical organization is/was also meant to tie into a hierarchical logging & monitoring system, that would for example allow you to set the loglevel, give a timing summary, or tee log messages to a specific file for a certain group.
 - Same thing for file output of passes. The prefix for output files can be set to use / for hierarchy separators, so the potentially very many output files end up in a hierarchical folder structure that makes sense to the user.
 - Pass groups are used as the bodies of loops and conditional passes, though nothing makes use of that yet.

Before we decided to remove the many CC-light-specific passes, I was intending to group the passes that were previously hardcoded as a group in backend in the default configuration. I intended to do the same with the mapper (make subpasses for initial placement, heuristic routing, and the post-route decomposition thing), but tearing those components apart was more difficult than I thought, so I ended up giving up on it. So, *right now*, indeed, I don't see a practical use case.

Even still, I don't understand why you'd then remove a feature that's already been built and documented. Sure, if something breaks or is otherwise in the way and no one is using it, get rid of it, I guess. But this hasn't even seen the light of day yet.

In general I find the "I don't see a use case" argument moot, because someone else might. You even found yourself at the other end of the argument with QI and the mapper. Why *not* present the user with an option to enable it (if indeed it is simple to implement)? Because the others "didn't see a use case." It feels so counterproductive. Isn't the end goal to make a tool, something that's useful *in general*? Not for ourselves, but for our users, so if there's any scenario where something might be helpful to someone now or in the future, why *not* leave it in? An elegant interface is one where everything has been reduced to the basic building blocks (perhaps with shorthand notation for commonly-used things), but with as many of those building blocks as possible.

If you feel strongly about this though, well, deleting stuff is easy...

> Wildcards are allowed in set_option but only (?) to select subpasses, seemingly not as prefix of the path

I'm not sure what you mean by this exactly. But the notation aims to follow the same semantics of file globs (when thinking of passes as directories). A `*` selects all subdirectories (subpasses) within a directory (group), and (with some implementation constraints) `**` matches one or more hierarchy levels at once. Some random examples:

```
root                y.b.v.option    y.*.option    y.**.option   **.option
 |- x                    -              -             -             X
 |- y (group)
 |  |- a                 -              X             X             X
 |  |- b (group)
 |  |  |- u              -              -             X             X
 |  |  |- v              X              -             X             X
 |  |  '- w              -              -             X             X
 |  '- c                 -              X             X             X
 '- z                    -              -             -             X
```

The only differences with glob are that directories (passes) *either* have subdirectories (subpasses) *or* files (settable options), and globbing the names of the files (options) makes no sense, so those cases aren't handled. Also, `y.**.v.option` would logically select only `y.b.v`, but currently throws an exception, because `**` is only supported as the last non-option element, because something made that annoying to implement, so for the sake of efficiency I didn't yet.

`set_option_recursively` just does `set_option` with a `**.` prefix. It's meant to be a less-cryptic alternative for something you might do relatively often, like `c.set_option_recursively('output_prefix', 'some/output/folder/%N')` to change the output directory for all passes at once.

One thing I do feel is missing here still is a way to select options based on pass type, so for example to set an option only for all clifford optimizers in the pass list, even if a similarly named option might exist for other pass types at some point as well. This also just didn't make the cut because I was trying to restrict myself to easy-to-implement stuff.

---

Note finally that the user isn't supposed to read the API reference first or even at all, there's supposed to be a user manual in front of it that guides them through basic usage. I'll concede that the API class documentation is very sparse right now and could be expanded, and that there are a lot of functions in the API reference that may not be useful right now, but just keep in mind that an API reference doesn't aim to teach someone how to use the corresponding library, in the same way that a dictionary doesn't aim to teach someone how to speak the corresponding language.

jvansomeren commented:
I am confused about documentation.

What is a user, what is a developer?

What is a user?
OpenQL is a compiler. So a user just uses it to compile a quantum program. But in doing so can set/unset options, add/delete passes, anything that the API allows. But also, the user must be able to understand what these mean, what a pass does and what the effect is when he adds/deletes passes. Of that the fundament is the IR. Therefore, the user documentation must also explain the structure and meaning of this IR, not only as external human readable thing but also as internal thing, because the external readable IR adds syntax and such which obscure the internal IR. This internal IR need not be described in terms of its actual implementation because that adds C++ details which again may obscure the meaning and clarity of the internal IR: it is about meaning, not syntax.
 
Anything not in the API is the domain of the developer: writing a new pass, extending/updating the IR, changing/writing a configuration file, etc.

Now, a user as described above may be able to do more than a 'simple user' may want to know. So one may consider a 'simple user' to be one who does not add/delete passes, who relies on internal defaults of pass order, who uses OpenQL out of the box, and therefore does not need to understand the internal IR to the depth as required for the 'advanced user'. For the 'simple user' it suffices to describe the IR in terms of what gets into OpenQL and what comes out. So here the cQASM documentation may be a start, supplemented by the variations that it supports (with/without mapping, with/without timing, etc.).

I propose to explain more clearly the target audience and the relation of the chapters of documentation to those kinds of audience in 'Overview', at the start of documentation.

jvansomeren commented:
get_custom_instructions => dump_custom_instructions
gate(... condstring:str) -> none: without condregs? This API can be taken out.
measure(...b0:Int)->none: specifies b0 to be float in the description

jvanstraten commented:
When I say "user," I'm talking about what you describe as a "simple user." Everyone in DiCarloLab except Wouter falls into that category, so I'd say it's the majority of our users (although I don't actually know how many people work in DiCarloLab). The Quantum Inspire people as well as entirely new users, no matter their intention, would also fall into this category. IMO they need the following things, and *only* the following things, in "OpenQL basics":

 - how to get OpenQL from PyPI (not bogged down by compilation instructions);
 - in a rough sense what a Platform is, how to make one for what they want to compile for, or if they don't know yet, what a sane default would be (rather than having to pick one with a whole pile of deprecated/unused garbage in it from the tests directory);
 - how to build a Program to compile for, either via cQASM or via the API (without first needing to read half a paper's worth on documentation about an IR, when they probably don't even know what that means yet -- I don't need to know what GCC's IR looks like to write a program for it or know approximately what -O0, -O3, -Os, -g, -ffast-math, and -flto do for me);
 - how to call Program.compile() and how to use (common) global or preferably pass options to get something useful out of it (we already have that in the tutorials section, but something including the visualizer would also be a nice way to introduce the existence of the Compiler interface); and last but not least
 - where they can find more information.

This part is meant to be "read like a book;" it should be a tutorial that the user can follow to get a basic idea of what OpenQL is, how to get it installed, and how to get a baseline Python file up and running that works on their machine. But that also means it needs to be short and interesting for everyone; if you start with "you first have to know X and Y and Z," especially when that isn't necessarily true for what a particular kind of user wants out of OpenQL, no one will actually read it. The "more information" part, on the other hand, should not read like a book, but act as additional information for parts referred to in basics, should a particular user decide for themselves that they want to know more about a particular subject. No one is supposed to read everything in a reference.

In my mind, there should be roughly three sections (with target audiences):

 - OpenQL basics (everyone): what I described above, culminating in the QX/DQCsim tutorials, and maybe new ones for the visualizer and statistics writer.
 - Reference ("advanced users" and contributors that touch the API or need to know how not to): anything else that's simple/advanced user-facing. I don't really like that the idea that this needs to include IR documentation, but if it's needed I'd put it in a section here.
 - Developer documentation (people contributing C++ code to OpenQL's codebase): what (new) developers need to know in addition. Perhaps this should be renamed to "contributor documentation."

The latter is mostly meta stuff for the project right now, but I was intending to stick the really in-depth things from what's currently in OpenQL basics about how the implementation works either in there or in comments inside the code. The not as in-depth things should go in the pass documentation (so its source is inside the code, it can be queried from within an interactive Python console, and pops up in the "Supported passes" section of the reference). In the basics section the mapper, scheduler, optimizers, etc should be introduced, but with no more detail than what you put in your OpenQL introduction slides.

The target audience etc. should *definitely* be described in the overview/introduction! No argument from me there. But I'd always only write that after having written everything else already.

Finally, you mention pointing basic users to the cQASM documentation, which has me very confused. That is very much reference information, and information that only works with context at that, since every "user" of cQASM (as in piece of software that includes libqasm) has its own way of interpreting the semantic tree generated by it. It's absolutely not libqasm's place to document that OpenQL discards timing information when you do basically anything except code generation, for instance (libqasm happens to be used by OpenQL, but it's certainly not a part of it). You'd *first* have to know that we're using it as an impromptu compiler IR and what that means before you can even begin to understand anything in that documentation and actually arrive at correct conclusions. The same thing applies for Quantum Inspire; I see the libqasm documentation more as documentation for the QI people who have to update their frontend and then have to update their documentation, than as something for the end user. Just like how C++ users don't tend to read the C++ language specificationl heck, it's not even publicly available.

So, the basics section should introduce cQASM *as OpenQL uses it* at some point in a clear and concise way, very clearly stating which subset OpenQL supports at the input and which subset it generates at the output, with a link to libqasm's documentation for further details at the end. There should also be a note there that the cQASM generated by OpenQL is not always actually valid cQASM, because it can't be right now (since cQASM 1.x does not support control-flow, and I'm 100% sure that the generator doesn't generate valid 1.1 syntax for classical constructs either yet, though it could probably be modified to do so), there is the whole skip/wait debacle that still needs to be dealt with, it's going to use custom gate names rather than only cQASM's default gateset, the operand lists may be different (measure insns getting their bit result register made explicit, crk being converted to cr, the angle op of cr not even being printed because it's not in the list of insns with an angle) etc., and we don't just want to throw an error when unrepresentable IR constructs are used, because it also serves as debug output.

jvanstraten commented:
I'm going to be AFK (driving) for the rest of the day, so I'll respond again for new stuff that pops up tomorrow.

jvansomeren commented:
In this light we might add to the API functions an annotation "for advanced users only",
after having defined/introduced "advanced user" in the introduction, or, more preferably, to the classes of those APIs. Because "user" == "people using API" is not true anymore otherwise.
Note that Miguel et al explicitly asked for modularity because they wanted control over which passes would be included in the compiler; when they want (and must be) able to take such decisions independent of us, they must know of IR attributes such as "with mapping" and "with timing".
Yes, and with cQASM I meant just enough to see the match between quantum gates in literature and those of OpenQL, not syntax, comment conventions, etc.

jvansomeren commented:
I wrote:

> I propose to undocument pass groups or to take them out of the implementation until we have a clear use case for them

I don't mean to discard and forget them.
From my CoSy background I see several missed opportunities (or didn't see them yet) and don't want to get into a backward compatibility issue when we would want to copy some of CoSy's engine design into OpenQL's pass group and option design.
The pass groups are about 80% CoSy's engine classes and some simple additions could turn that into 120%; that +20% because we encountered deficiencies in CoSy as well in the end such as lack of dynamics wrt pass management that are already in OpenQL.
So the easiest would be to keep them out of the documented interfaces.
And then I would like to:

- explain the design of CoSy's engines, engine classes, option translation and logging
- discuss how (and whether) that functionality could be adopted by OpenQL
- followed by updating pass group documentation/implementation

This requires a face2face meeting, to save time.

jvansomeren commented:
Regarding `Topology::sort_neighbors_by_angle`, it belongs to Topology in my opinion. But it is a thing that is not mature yet. It has to do with subsetting the qubits. So could be used to define cores or particular sections of qubits with particular functionalities. In this case, the section contains those that are on the shortest path between two given qubits. But a section may have attributes like being convex (any shortest path between two qubits in the section only contains qubits in the section), being connected.
In the current case, the space of connections is planar, i.e. connections only cross in qubits (nodes).
With non-planar grids, the spaces become higher-dimensional.
Fully connected grids have the highest dimension.
Reading this, I fear that some of the ibm grids are non-planar and that thus this sorting doesn't work for those grids, i.e. we have a bug.

jvansomeren commented:
The cc_light default config file has a wrong cnot decomposition:
```
"gate_decomposition": {
        "cnot %0,%1" : ["ry90 %0","cz %0,%1","ry90 %1"]
    }
```
It should be (note the `ry90 %0` to `ym90 %1` gate change):
```
"gate_decomposition": {
        "cnot %0,%1" : ["ym90 %1","cz %0,%1","ry90 %1"]
    }

```

jvansomeren commented:
Option generate_code is still used by the multi-core branch. When no, the final code generation is not done.
Option mappathselect with value borders doesn't follow the borders of the chip but of the search space provided that the chip's coupling graph is planar.
Option maptiebreak 's first/last values assume that the coupling graph is planar and neighbors were sorted; and assume knowledge on how the alternatives are generated then (first is left-most alternative with 2q near target, last is right-most alternative with 2q near source).
Option use_default_gates: TODO must be moved to TODO section of documentation.
Option write_qasm_files: end of sentence is incorrect.

jvansomeren commented:
With the visualizer, radius etc is in pixels. Isn't a pixel a variable thing, i.e. can it be that the configuration needs to be updated for different pixel density?

jvansomeren commented:
With the pass cQASM reader, the uninitiated reader doesn't understand the relation between lib qasm and cQASM reader.

jvansomeren commented:
Several options have a file name string as value: `cqasm_file`, `gateset_file`, `output_prefix`, and probably several more. It is not clear how these file name strings are interpreted: are these absolute path names, are these relative to the installation root, are these relative to the current directory? Is there something to control this?

jvansomeren commented:
Mapper pass: I would like to discuss what it does and how it can be improved. The `lookahead_mode` is troubled by the interaction with recursion and that only subsets of gates (2q gates, nonNN 2q gates) are relevant in lookahead; of course 1q gates and NN 2q gates can be immediately mapped. Perhaps the recursion should be taken out of the documentation and the `lookahead_mode` be reduced to a mode without recursion.

Without recursion, the `lookahead_mode` option implies:
`no`: just map the gates in the order of the circuit; other modes assume a dependence graph and avlist; a 1q gate or a NN 2q gate maps trivially; map a nonNN 2q gate by generating alternatives for this nonNN 2q gate and selecting the best
`1qfirst`: map the 1q gates before any 2q gates in avlist until only 2q gates remain; then map one, i.e. the most critical 2q gate by generating alternatives for this nonNN 2q gate and selecting the best
`noroutingfirst`: map the 1q gates and NN 2q gates in avlist until only nonNN 2q gates remain; then map one, i.e. the most critical nonNN 2q gate by generating alternatives for this nonNN 2q gate and selecting the best
`all`: map the 1q gates and NN 2q gates in avlist until only nonNN 2q gates remain; then generate alternatives for all nonNN 2q gates and map that nonNN 2q gate corresponding to the best alternative selected.

`path_selection_mode`: not chip border but search space border, and only if coupling graph is planar

`recurse_on_nn_two_qubit`: this is the option interfering during recursion with lookahead_mode; this one and the following recursion options could be suppressed from documentation.

`reverse_swap_if_better`: its implementation may even be exactly wrong; and it depends on swap decomposition on the particular target, so perhaps it should be made more clear in which cases it does something and what it does then, also to make it independent of the target

`tie_break_method`: first and last depend on knowledge of alternative path generation; and taking the first/last is then only deterministic when the coupling graph is planar

jvansomeren commented:
Resources and the new way of specifying resource constraints, reading the documentation.

This made the resource specification more powerful: more types of resources can be described.

Regarding Instrument resource, does the detuned_qubits resource match the description of general operation in the Instrument resource section? A mw and a flux don't need to have the same duration for running in parallel, nor need to start in the same cycle. The instrument in this case is the particular frequency choice for a qubit: default or detuned. Mw needs default; some flux need detuned on a qubit defined by the flux operand pair. Qubit exclusive use is assumed (other instrument).
The section reads as an introduction to resources and is very useful in this respect, so should be kept but must be made accurate. I propose to add that this is a rationale/example/motivation, not a specification; or to add other types of resources (such as detuned_qubits) that are somehow different must must be supported as well. Or that the `if and only if` before the first bullet list is updated.

Key here is `they use the same instrument function;`, not the other ones.
For a 1q mw and a 2q flux to be in parallel regarding the detuned_qubits resource, they must not both use the same instrument object. In the detuned_qubits example config, the 6 objects model the 6 qubits where a conflict could arise. The 1q and 2q select the instrument object thru different size-specialized predicates.
So `same instrument function` means: `same instrument (object) in the set of a particular instrument type`.

Regarding the configuration structure: a required condition for an incoming gate to match a predicate is:
> its instruction set definition object has values for all keys specified;
When one forgets to add a particular key in a gate specification (e.g. "mw" as "type" in "x45"), it will never be matched; could this go unnoticed somehow? If so, consider checking this somehow. I don't have an idea now.

What could be emphasized more is that there always are two gates for which one wants to know whether they can be in parallel. For each such pair one checks all resource types. When for one resource type, both gates use (select to/match) the same instruction object, they cannot be in parallel. This `operational` definition is somehow assumed.

I don't understand `"exclusive"`. Or, I don't understand the function section. Or, I don't understand: `a list of gate keys as specified in the structure above`

The original resource specs didn't support three-or-more-qubit gates because the underlying architectures don't support them. So why is their support now in?

Perhaps suppress the multi-core channel resource since it is not mature yet. There are many implicits here (deriving core index from qubit index, full core connectivity) which need ability to be specified. The documentation here is not ready yet either. The channel resource would only be allocated by inter-core-gates, a subset of 2q gates, each such gate allocates two channels, one in each core. I don't understand: `This is mostly for compatibility with the original channel resource, which didn’t check for this.`

jvansomeren commented:
With the Release procedure, I am missing the writing of Release Notes. This is a summary of what has changed, what is new, what is incompatible and gives recipes how to overcome the latter. It also summarizes what is deprecated from now on and will be removed in a later version. It need not be a lot of text but should refer to the necessary texts in the documentation; it is an entry-level document.

jvansomeren commented:
In C++ coding conventions, the section Namespaces appears twice.
Regarding virgin constructors and an init method, I'm still curious why this was (or why I thought this was) needed in the mapper.

jvansomeren commented:
I concluded with reviewing the PR documentation (as generated from within the docs directory).

jvanstraten commented:
Up to https://github.com/QE-Lab/OpenQL/pull/405#issuecomment-831175809:

> In Python API, compilation strategy is mentioned, a default compilation strategy which not be good enough and that you can change it. Only in class Platform it is described (with the compiler config) while it should be described in class Compiler at least.

Lazily added a single sentence description of what it's for (bunched in with 94a69be). This kind of thing is more something for user-manual style docs than API docs anyway.

> (discussion about pass groups)

Removed all traces of pass groups that I'm aware of from the API based on `QL_HIERARCHICAL_PASS_MANAGEMENT` in `config.h.template`. I didn't want to remove everything completely, because I had to change a lot of docstrings to make everything sane again.

> get_custom_instructions => dump_custom_instructions

This is an old API, not one that I added, so for backward compat it should stay in. I've added a dump_ synonym for it and deprecated the get_ version though.

> gate(... condstring:str) -> none: without condregs? This API can be taken out.

It's for the always/never variants, and mirrors the internal function signature. Leaving it in allows gates to be added procedurally using `gate(... condstring, *condregs)` regardless of condition type (or lack thereof). Also, it's rather annoying to take it out because it's generated from C++'s optional argument system.

Also fixed a few compilation instruction things found by Quinten.

jvanstraten commented:
Up to https://github.com/QE-Lab/OpenQL/pull/405#issuecomment-832481890:

> (sort_neighbors_by_angle stuff)

Replaced the FIXME with a TODO note referring to this comment, don't know what else to do with it for now.

> (CC-light CNOT gate decomposition)

Fixed, but note this changed golden files. It wasn't broken during refactoring.

> Option generate_code is still used by the multi-core branch. When no, the final code generation is not done.

It's defunct because the QISA code generation pass was removed, so regardless of whether it's yes or no, final code generation is not done.

> Option mappathselect with value borders doesn't follow the borders of the chip but of the search space provided that the chip's coupling graph is planar.

Updated.

> Option maptiebreak 's first/last values assume that the coupling graph is planar and neighbors were sorted; and assume knowledge on how the alternatives are generated then (first is left-most alternative with 2q near target, last is right-most alternative with 2q near source).

Updated.

> Option use_default_gates: TODO must be moved to TODO section of documentation.

I'm not sure what you mean by this, and in general the option documentation is generated, so it's not really possible to add RST directives and whatnot to them without making everything vastly more complicated.

> Option write_qasm_files: end of sentence is incorrect.

Fixed.

> With the visualizer, radius etc is in pixels. Isn't a pixel a variable thing, i.e. can it be that the configuration needs to be updated for different pixel density?

I highly doubt that the CImg library is aware of pixel density and GUI scale, and if not, it's not really something that can easily be fixed. Internally CImg just makes a bitmap image (which it can also write to a file as such) which logically uses pixel units, and eventually makes a window that displays this bitmap. It's no more intelligent than that. If this is/becomes a problem maybe something can be tacked on to fix it later, but that'd be a feature request that really shouldn't be bunched up in a refactoring commit.

> With the pass cQASM reader, the uninitiated reader doesn't understand the relation between lib qasm and cQASM reader.

Added short explanation of relationship between the two, with reference to libqasm's ReadTheDocs page.

> (filename relativeness)

Anywhere this is not specified, filenames must either be absolute or relative to the current working directory as tracked by the OS. IMO this is just how computers work, so I really don't feel like this is something worth repeating everywhere...

We could quite easily change the behavior since all file I/O is wrapped by `utils`, but I don't really see why in this case? I suppose I could see it be useful when these option values are specified in a compiler configuration file, but any implementation for that would have to be really contrived, because it would mean there are special cases for certain options when they're specified in a certain way...

jvanstraten commented:
For https://github.com/QE-Lab/OpenQL/pull/405#issuecomment-832506285:

> Mapper pass: I would like to discuss what it does and how it can be improved. [...]

Changing the functional implementation of the mapper should be a separate issue/PR, especially at this point in the process.

> Without recursion, the `lookahead_mode` option implies: [...]

Updated option docs accordingly.

> `path_selection_mode`: not chip border but search space border, and only if coupling graph is planar

Already updated along with the global option.

> `recurse_on_nn_two_qubit`: this is the option interfering during recursion with `lookahead_mode`; this one and the following recursion options could be suppressed from documentation.

As I mentioned in an earlier comment, the option documentation (including pass option documentation) is generated from the options exactly as they exist in the sources and thus include everything, and I firmly believe this is a good thing. That's what it's a reference for. So, instead of surpressing, I added notes stating that these are advanced and unstable options, for `recurse_on_nn_two_qubit` including that it affects `lookahead_mode` in a currently-undocumented way.

> `reverse_swap_if_better`: its implementation may even be exactly wrong; and it depends on swap decomposition on the particular target, so perhaps it should be made more clear in which cases it does something and what it does then, also to make it independent of the target

I feel like the docstring I added to it already explained this, but I expanded upon it to make it more clear. IMO it's up to the user to define the composition such that it complies with the convention.

I also more properly documented the general behavior of the mapper, and noted there what the requirements on the gate/decomposition rule definitions in the platform are.

jvanstraten commented:
> With the Release procedure, I am missing the writing of Release Notes.

It was very vaguely in there (though I called it a short changelog), but I expanded on it now and added a few other missing steps while I was at it. I was intending to update it while doing the actual release, so I hadn't looked at it much yet.

> In C++ coding conventions, the section Namespaces appears twice.

Fixed.

> Regarding virgin constructors and an init method, I'm still curious why this was (or why I thought this was) needed in the mapper.

Just because C++ is annoying with its member initialization syntax. There are ways around it, but it quickly becomes a lot more messy than it's worth. It's easier to just wrap a class member in `utils::Opt` or `utils::Ptr` when you can't easily initialize it before the constructor's function body; this lets you do everything that virgin constructor + init lets you do, but ensures that you can never use the object accidentally before initializing it (you'll get a runtime error if you do, stating that you're dereferencing an empty `Opt` or `Ptr`) and still lets you use RAII for the affected class where this issue doesn't exist.

I believe that's everything except for the functional limitations for the instruments and the still-missing user docs and tests for the new compiler API.
